client/Lykke.AlgoStore.MatchingEngineAdapter.Client/RequestManager.cs

```diff
@@ -34,58 +33,44 @@ public void SetClientAndInstanceId(string clientId, string instanceId)
             _meaCommunicator.Start();
         }
 
-        public (WaitHandle, uint) MakeRequest<T>(MeaRequestType requestType, T message)
+        public Task<IMessageInfo> MakeRequestAsync<T>(MeaRequestType requestType, T message)
         {
-            var manualResetEvent = new ManualResetEvent(false);
+            var taskCompletionSource = new TaskCompletionSource<IMessageInfo>();
             var nextRequestId = GetNextRequestId();
 
-            _waitHandles.TryAdd(nextRequestId, manualResetEvent);
-            _meaCommunicator.SendRequest(nextRequestId, (byte)requestType, message);
+            _queuedMessages.TryAdd(nextRequestId, taskCompletionSource);
+            // Wait here because otherwise we can't return the task from the completion source
+            _meaCommunicator.SendRequestAsync(nextRequestId, (byte)requestType, message).Wait();
 
-            return (manualResetEvent, nextRequestId);
-        }
-
-        public object GetResponse(uint requestId)
-        {
-            object response;
-
-            if (!_responses.TryRemove(requestId, out response))
-                throw new KeyNotFoundException($"A message with the ID {requestId} does not exist");
-
-            return response;
+            return taskCompletionSource.Task;
         }
 
         private void ProcessResponse(IMessageInfo messageInfo)
         {
             if (messageInfo == null)
                 throw new ArgumentNullException(nameof(messageInfo));
 
-            if (!_waitHandles.ContainsKey(messageInfo.Id))
+            if (!_queuedMessages.ContainsKey(messageInfo.Id))
                 return;
 
-            ManualResetEvent manualResetEvent;
+            TaskCompletionSource<IMessageInfo> taskCompletionSource;
 
-            if (!_waitHandles.TryRemove(messageInfo.Id, out manualResetEvent))
+            if (!_queuedMessages.TryRemove(messageInfo.Id, out taskCompletionSource))
                 return;
 
-            _responses.TryAdd(messageInfo.Id, messageInfo.Message);
-            manualResetEvent.Set();
+            taskCompletionSource.SetResult(messageInfo);
         }
 
         private void SendAuthRequest()
         {
             var request = new PingRequest { Message = $"{_clientId}_{_instanceId}" };
 
-            (var waitHandle, var requestId) = MakeRequest(MeaRequestType.Ping, request);
+            var task = MakeRequestAsync(MeaRequestType.Ping, request);
 
-            ThreadPool.QueueUserWorkItem((state) =>
+            task.ContinueWith((t) =>
             {
-                waitHandle.WaitOne();
-
-                var response = GetResponse(requestId) as PingRequest;
+                var response = t.Result.Message as PingRequest;
 
-                // This will cause an unhandled exception and exit/crash the algo instance
-                // It is unlikely that an honest instance will get into a situation where it can't authenticate
                 if (response.Message == "Fail")
```

Comment:
The cast should never fail here as that's the expected reply to the message. If it does, that would mean something has gone horribly wrong and it should probably throw anyway.

`response?.Message` wouldn't help because in the case response is null, the comparison will amount to `null == "Fail"` which would be false, thereby circumventing the authentication check.